uaclient/tests/test_auto_attach.py

```diff
@@ -0,0 +1,132 @@
+import contextlib
+import io
+
+import mock
+import pytest
+
+from lib.auto_attach import check_cloudinit_userdata_for_ua_info, main
+from uaclient.exceptions import (
+    AlreadyAttachedOnPROError,
+    ProcessExecutionError,
+)
+
+
+class TestCheckCloudinitUserdataForUAInfo:
+    @mock.patch("lib.auto_attach.which")
+    def test_check_cloudinit_data_returns_false_if_no_cloudinit(self, m_which):
+        m_which.return_value = False
+        assert False is check_cloudinit_userdata_for_ua_info()
+
+    @mock.patch("lib.auto_attach.subp")
+    @mock.patch("lib.auto_attach.which")
+    def test_check_cloudinit_data_returns_false_if_query_error(
+        self, m_which, m_subp
+    ):
+        m_which.return_value = True
+        m_subp.side_effect = ProcessExecutionError("test")
+
+        assert False is check_cloudinit_userdata_for_ua_info()
+
+    @mock.patch("lib.auto_attach.subp")
+    @mock.patch("lib.auto_attach.which")
+    def test_check_cloudinit_data_returns_false_if_yaml_error(
+        self, m_which, m_subp
+    ):
+        m_which.return_value = True
+        m_subp.return_value = (
+            """\
+            [invalid, yaml]:
+              - test
+            """,
+            "",
+        )
+
+        assert False is check_cloudinit_userdata_for_ua_info()
+
+    @pytest.mark.parametrize(
+        "expected,userdata",
+        (
+            (
+                False,
+                "",
+            ),
+            (
+                False,
+                """\
+                #cloud-config
+                runcmd:
+                  - echo "test"
+                """,
+            ),
+            (
+                True,
+                """\
+                #cloud-config
+                ubuntu_advantage:
+                 - echo "test"
+                """,
+            ),
+        ),
+    )
+    @mock.patch("lib.auto_attach.subp")
+    @mock.patch("lib.auto_attach.which")
+    def test_check_cloudinit_data(self, m_which, m_subp, expected, userdata):
+        m_which.return_value = True
+        m_subp.return_value = (userdata, "")
+
+        assert expected is check_cloudinit_userdata_for_ua_info()
+
+
+class TestMain:
+    @pytest.mark.parametrize(
+        "ubuntu_advantage_in_userdata",
```

Comment:
Nit: Not sure if you use this layout to make it extensible in the future. But, this could be simplified to:

```py
@pytest.mark.parametrize("ubuntu_advantage_in_userdata", [True, False])
```